File: frontend/lib/models/wishlist_item.dart
import 'category.dart';

enum ItemStatus { WANTED, DONE }

class WishlistItem {
  final String? id;
  final String title;
  final String? description;
  final String? notes;
  final String? imageUrl;
  final String? externalUrl;
  final String? linkPreviewImageUrl;
  final String? linkPreviewTitle;
  final ItemStatus status;
  final Category category;
  final DateTime? createdAt;

  const WishlistItem({
    this.id,
    required this.title,
    this.description,
    this.notes,
    this.imageUrl,
    this.externalUrl,
    this.linkPreviewImageUrl,
    this.linkPreviewTitle,
    this.status = ItemStatus.WANTED,
    required this.category,
    this.createdAt,
  });

  factory WishlistItem.fromJson(Map<String, dynamic> json) => WishlistItem(
        id: json['id'],
        title: json['title'],
        description: json['description'],
        notes: json['notes'],
        imageUrl: json['imageUrl'],
        externalUrl: json['externalUrl'],
        linkPreviewImageUrl: json['linkPreviewImageUrl'],
        linkPreviewTitle: json['linkPreviewTitle'],
        status: json['status'] == 'DONE' ? ItemStatus.DONE : ItemStatus.WANTED,
        category: Category.fromJson(json['category']),
        createdAt: json['createdAt'] != null
            ? DateTime.tryParse(json['createdAt'])
            : null,
      );

  Map<String, dynamic> toJson() => {
        'title': title,
        'description': description,
        'notes': notes,
        'imageUrl': imageUrl,
        'externalUrl': externalUrl,
        'status': status.name,
        'category': {'id': category.id},
      };

  WishlistItem copyWith({ItemStatus? status}) => WishlistItem(
        id: id,
        title: title,
        description: description,
        notes: notes,
        imageUrl: imageUrl,
        externalUrl: externalUrl,
        linkPreviewImageUrl: linkPreviewImageUrl,
        linkPreviewTitle: linkPreviewTitle,
        status: status ?? this.status,
        category: category,
        createdAt: createdAt,
      );
}

File: src/main/java/com/mangetout/model/WishlistItem.java
package com.mangetout.model;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;

@Document(collection = "wishlist_items")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class WishlistItem {

    @Id
    private String id;

    @NotBlank
    private String title;

    private String description;
    private String notes;
    private String imageUrl;
    private String externalUrl;
    private String linkPreviewImageUrl;
    private String linkPreviewTitle;

    private ItemStatus status = ItemStatus.WANTED;

    // Category is embedded directly — idiomatic MongoDB (no join needed on read)
    @NotNull
    private Category category;

    // References the Couple this item belongs to — used for scoping all queries
    private String coupleId;

    // Store only id + username instead of embedding the full User document
    private String addedById;
    private String addedByUsername;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;
}
